### ISSUE TITLE
Event driven system as SUT

### ISSUE THREAD

**Issue by bergtwvd** [mukatee/osmo#60]
I am looking at various MBT tools for testing event driven systems.

I can stimulate the SUT via the OSMO test model, but how can I trigger a guard to become true on events received from the SUT? I can update the OSMO test model on a callback from the SUT, but is this sufficient for OSMO to know that a certain guards has evaluated to true?  Is there a specific `update` method I should call to trigger OSMO?
Thanks!

**Comment by mukatee** [comment]
Hello,

There is no specific need to call any update function or such. As you describe your use case, I assume you are doing online model-based testing.Meaning test generation stimulates the SUT with some event, and there is some callback from the SUT back to the tester application/model. Just update the model state when you receive the callback. OSMO calls the guard statements at each step, so as long as your model state is up-to-date, it will be reflected in the guards.

If I misunderstand your intent, maybe you can do some simple toy example to illustrate your problem case?

Teemu

**Comment by bergtwvd** [comment]
Hi, 

Yes, that's right, it is about online testing. I constructed a simple example (see attached zip) of a vending machine that runs as a separate SUT, using an adapter to access it. As such there are no issues, just a few points I noticed.

While constructing the example, I noticed that OSMO deliberately generates an error if there are not Test Steps to perform. This situation may be temporal, since the SUT may respond with an event (in the future) that will enable one of the test steps. Something to be aware off while creating the test model. I added for example a wait step.

I may need to create a test model with parallel states (hence multiple instances of OSMO running on different threads). It that possible?

I assume OSMO does a kind of busy wait loop, constantly checking with guard conditions evaluate to true. Correct?

[src.zip](https://github.com/mukatee/osmo/files/2923842/src.zip)

**Comment by mukatee** [comment]
Yes you are correct, OSMO basically loops through the model all the time and looks for enabled steps. One of the enabled steps is then taken for execution at each iteration. 

Currently it is built with the assumption that there should always be a way forward, and hitting a state where there is not path forward is considered by the generator as an anomaly, and an "error" state in the model. But I see it is not necessarily so in your case. A specific wait step should fix that as you have noted yourself. It is a very good point and approach for that type of testing.

For parallel states, it depends on what you want to achieve with those. You can of course start multiple OSMO instances in parallel, but they will not be aware of each other. If each of them simulates a client on its own, there should be no issue. If you need shared state, that would require you to build some way for the models to communicate. Like an in-memory data cache/store. Again, depends how deep you want to go.

Of course, you can also build parallel states within a single model, where you just select which state is updated based on some value you receive/give when interacting with the SUT. Keeping it simple in this type scenario is maybe the biggest challenge, but sounds very interesting :)
